# Have eyes on the data


In [3]:
# importing data
import pandas as pd
from textblob import TextBlob
import numpy as np

In [23]:
# published news data.

ratings = pd.read_csv("/home/nesrex/My files/2 Work /Kifya AI Mastery/week o1/Data/raw_analyst_ratings.csv")

In [24]:
ratings.head(5)

Unnamed: 0                                           headline  \
0           0            Stocks That Hit 52-Week Highs On Friday   
1           1         Stocks That Hit 52-Week Highs On Wednesday   
2           2                      71 Biggest Movers From Friday   
3           3       46 Stocks Moving In Friday's Mid-Day Session   
4           4  B of A Securities Maintains Neutral on Agilent...   

                                                 url          publisher  \
0  https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1  https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2  https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3  https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4  https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   

                        date stock  
0  2020-06-05 10:30:54-04:00     A  
1  2020-06-03 10:45:20-04:00     A  
2  2020-05-26 04:30:07-04:00     A  
3  2020-05-22 12:45:06-04:00     A  
4  2020-05-22 11:38:59-04:00     A

In [6]:
ratings["stock"].unique()

AttributeError: module 'numpy' has no attribute 'index_exp'

In [7]:
ratings.shape

(1407328, 6)

In [8]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1407328 entries, 0 to 1407327
Data columns (total 6 columns):
 #   Column      Non-Null Count    Dtype 
---  ------      --------------    ----- 
 0   Unnamed: 0  1407328 non-null  int64 
 1   headline    1407328 non-null  object
 2   url         1407328 non-null  object
 3   publisher   1407328 non-null  object
 4   date        1407328 non-null  object
 5   stock       1407328 non-null  object
dtypes: int64(1), object(5)
memory usage: 64.4+ MB


In [9]:
# headline length

ratings['word_count'] = (ratings['headline']
    .str.strip()
    .str.split()
    .str.len()
    .fillna(0)
)
ratings.head(20)

AttributeError: module 'numpy' has no attribute 'matrix'

In [9]:
published_nums = ratings['publisher'].value_counts(dropna = False).reset_index()
published_nums.columns = ['publisher', 'total publication']

published_nums

publisher  total publication
0         Paul Quintaro             228373
1            Lisa Levin             186979
2     Benzinga Newsdesk             150484
3         Charles Gross              96732
4         Monica Gerson              82380
...                 ...                ...
1029          MoneyGeek                  1
1030             muathe                  1
1031      Robert Morris                  1
1032    LeftCoastHedgie                  1
1033     Jeremie Capron                  1

[1034 rows x 2 columns]

So, the top three large publishers are:

1. Paul Quintaro	228373
2. Lisa Levin	186979
3. Benzinga Newsdesk 150484

In [18]:
# to see publications in a date and understand the publishing trend, we have to have the date separetly.

ratings['date_only'] = (
    ratings['date']
    .str.split()
    .str.get(0)
   )

daily_Post = ratings['date_only'].value_counts().reset_index()
daily_Post = daily_Post.sort_values('date_only')
daily_Post.tail(10)

date_only  count
1033  2020-06-02    544
202   2020-06-03    789
1078  2020-06-04    538
76    2020-06-05    932
3557  2020-06-06     11
3135  2020-06-07     25
236   2020-06-08    765
182   2020-06-09    804
178   2020-06-10    806
1036  2020-06-11    544

In [19]:
type(ratings['date_only'])

pandas.core.series.Series

## Text Analysis (Sentiment analysis of Headlines)

In [13]:
#data = pd.DataFrame({'headline': ["do not buy stocks", 'being in srock market is great', 'Stocks That Hit 52-Week Highs On Friday', 'Company faces backlash over environmental concerns', "Stock markets rally as investors cheer positive..."]})

def check_polarity(text):
    blob = TextBlob(text)
    polarity = blob.sentiment.polarity
    if polarity > 0:
        return("positive")
    elif polarity < 0:
        return('Negative')
    else:
        return('Neutral') 


ratings['sentiment'] = ratings['headline'].apply(check_polarity);

#ratings.head()

In [11]:
ratings.head()

Unnamed: 0                                           headline  \
0           0            Stocks That Hit 52-Week Highs On Friday   
1           1         Stocks That Hit 52-Week Highs On Wednesday   
2           2                      71 Biggest Movers From Friday   
3           3       46 Stocks Moving In Friday's Mid-Day Session   
4           4  B of A Securities Maintains Neutral on Agilent...   

                                                 url          publisher  \
0  https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1  https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2  https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3  https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4  https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   

                        date stock  
0  2020-06-05 10:30:54-04:00     A  
1  2020-06-03 10:45:20-04:00     A  
2  2020-05-26 04:30:07-04:00     A  
3  2020-05-22 12:45:06-04:00     A  
4  2020-05-22 11:38:59-04:00     A

### Time series analysis

In [20]:
ratings.head()

headline  \
0            Stocks That Hit 52-Week Highs On Friday   
1         Stocks That Hit 52-Week Highs On Wednesday   
2                      71 Biggest Movers From Friday   
3       46 Stocks Moving In Friday's Mid-Day Session   
4  B of A Securities Maintains Neutral on Agilent...   

                                                 url          publisher  \
0  https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1  https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2  https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3  https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4  https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   

                        date stock  
0  2020-06-05 10:30:54-04:00     A  
1  2020-06-03 10:45:20-04:00     A  
2  2020-05-26 04:30:07-04:00     A  
3  2020-05-22 12:45:06-04:00     A  
4  2020-05-22 11:38:59-04:00     A

In [25]:
# the date should be the index

ratings.set_index("date")
del(ratings['Unnamed: 0'])
ratings.head()

headline  \
0            Stocks That Hit 52-Week Highs On Friday   
1         Stocks That Hit 52-Week Highs On Wednesday   
2                      71 Biggest Movers From Friday   
3       46 Stocks Moving In Friday's Mid-Day Session   
4  B of A Securities Maintains Neutral on Agilent...   

                                                 url          publisher  \
0  https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1  https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2  https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3  https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4  https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   

                        date stock  
0  2020-06-05 10:30:54-04:00     A  
1  2020-06-03 10:45:20-04:00     A  
2  2020-05-26 04:30:07-04:00     A  
3  2020-05-22 12:45:06-04:00     A  
4  2020-05-22 11:38:59-04:00     A